Add cuDNN support to batch normalization function

bkvogel opened the pull request:
I have added cuDNN support to the batch normalization function and updated the corresponding link. This involved writing Python wrappers for the new cuDNN functions and using the cuDNN implementation if cuDNN is enabled. Since cuDNN updates the running average statistics (mean and variance), I also moved these computations from the link into the function.

okuta commented:
Why don’t you support float16 and float64?

bkvogel commented:
I disabled float16 because the tests failed using cuDNN. I see other functions such as relu.py allow it to run anyway if the cudnn version is at least 3000. I didn't realize I did not support float64. Could you give me an example of another function that handles all float types correctly so that I could use it as a template? Thanks.

bkvogel commented:
I have modified the tests to be similar to the newer batch normalization tests in master that check all float types. float16 and float64 types are tested now. I still do not allow the cuDNN version to use float16 since it fails the test on my machine using cuDNN v5. So if float16 is detected, cupy version is used instead. How does this look?

okuta commented on chainer/functions/normalization/batch_normalization.py:
```diff
@@ -4,20 +4,53 @@
 class BatchNormalizationFunction(function.Function):
 
-    def __init__(self, eps=1e-5):
+    def __init__(self, mean, var, train, decay=0.9, eps=2e-5, use_cudnn=True):
```
Please do not change argument order.

bkvogel replied:
OK, I will rewrite to keep "eps" as first default arg.

okuta commented on chainer/functions/normalization/batch_normalization.py:
```diff
@@ -27,100 +60,266 @@ def check_type_forward(self, in_types):
     def forward(self, inputs):
         xp = cuda.get_array_module(*inputs)
-        x, gamma, beta = inputs[:3]
+        x, gamma, beta = inputs
```
Do not delete forward function argument for backward compatibility.

bkvogel replied:
This function was never documented, though. Is it okay to remove the optional arguments as long as I do not brake any of the documented APIs?

okuta replied:
Sorry, This comment was very confusing.

We should not keep backward compatibility of undocumented functionality.
But, `fixed_batch_normalization` wants gradients of `mean` and `var`.
And the computational graph includes `mean` and `var` Variable in that case.

okuta commented on chainer/functions/normalization/batch_normalization.py:
```diff
@@ -27,100 +60,266 @@ def check_type_forward(self, in_types):
+            # into a 2-dim array with channels as second dim and m=<product
+            # of all dimensions except the 2nd dimension> as the first
+            # dimension.
+            self.cudnn_dim_ok = True
```
How about `self.cudnn_dim_ok = x.ndim == 2 or x.ndim == 4`?

okuta commented on chainer/functions/normalization/batch_normalization.py:
```diff
@@ -27,100 +60,266 @@ def check_type_forward(self, in_types):
             self.x_hat = x_mu / self.std[expander]
             y = gamma * self.x_hat
             y += beta
+        elif cuda.cudnn_enabled and self.use_cudnn and self.cudnn_dim_ok:
```
Please check cuDNN version. Chainer supports cuDNN from v2 to v5.

okuta commented on chainer/functions/normalization/batch_normalization.py:
```diff
@@ -27,100 +60,266 @@ def check_type_forward(self, in_types):
+            # computation runningMean
+            if self.mean_cache is None:
+                # Output cache to speed up bacward pass (recommended to enable)
+                self.mean_cache = xp.zeros_like(gamma)
```
How about using `empty_like`?

okuta commented on chainer/functions/normalization/batch_normalization.py:
```diff
@@ -27,100 +60,266 @@ def check_type_forward(self, in_types):
+                # Output cache to speed up bacward pass (recommended to enable)
+                self.mean_cache = xp.zeros_like(gamma)
+                # Output cache to speed up bacward pass (recommended to enable)
+                self.var_cache = xp.zeros_like(gamma)
```
How about using `empty_like`?

okuta commented on chainer/functions/normalization/batch_normalization.py:
```diff
@@ -27,100 +60,266 @@ def check_type_forward(self, in_types):
 
+        if x[0].dtype == numpy.float16:
+            # cuDNN v5 batch normalization does not seem to support float16.
+            self.use_cudnn = False
```
`self.use_cudnn` had assigned at `forward`.

okuta commented on chainer/functions/normalization/batch_normalization.py:
```diff
@@ -27,100 +60,266 @@ def check_type_forward(self, in_types):
 
 
-def batch_normalization(x, gamma, beta, eps=1e-5):
+def batch_normalization(x, gamma, beta, running_mean=None,
```
Do not change argument order for backward compatibility.

bkvogel replied:
Thank you for the comment. I was just made aware that we should try to maintain backward compatibility on any documented functions, even if they will probably only ever be used by the corresponding link. I will try to rewrite to maintain backward compatibility.

okuta commented on chainer/functions/normalization/batch_normalization.py:
```diff
@@ -129,28 +328,30 @@ def batch_normalization(x, gamma, beta, eps=1e-5):
 
 
-def fixed_batch_normalization(x, gamma, beta, mean, var, eps=1e-5):
+def fixed_batch_normalization(x, gamma, beta, fixed_mean, fixed_var, eps=2e-5):
```
Do not change arguments name for backward compatibility.

okuta commented on chainer/links/normalization/batch_normalization.py:
```diff
@@ -101,7 +100,8 @@ def __call__(self, x, test=False, finetune=False):
         """
         use_batch_mean = not test or finetune
-
+        print('use_batch_mean = ', use_batch_mean)
```
Please remove this line.

okuta commented on chainer/links/normalization/batch_normalization.py:
```diff
@@ -114,31 +114,23 @@ def __call__(self, x, test=False, finetune=False):
-                func.var *= (1 - decay) * adjust  # reuse buffer as a temporary
-                self.avg_var += func.var
-                del func.var
+            print('decay = ', decay)
```
Please remove this line.

okuta commented on tests/chainer_tests/functions_tests/normalization_tests/test_batch_normalization.py:
```diff
@@ -132,19 +140,22 @@ def test_forward_cpu(self):
     @attr.gpu
     @condition.retry(3)
     def test_forward_gpu(self):
+        print('In: test_forward_gpu()...')
```
Please remove this line.

okuta commented on cupy/cuda/cudnn.pyx:
```diff
@@ -585,6 +618,71 @@ cpdef poolingBackward(
+        <TensorDescriptor>bnScaleBiasMeanVarDesc, <void*>bnScale,
+        <void*>bnBias, <float>exponentialAverageFactor,
+        <void*>resultRunningMean, <void*>resultRunningVariance,
+        <float>epsilon, <void*>resultSaveMean, <void*>resultSaveInvVariance)
```
`<double>epsilon` is correct.

okuta commented on cupy/cuda/cudnn.pyx:
```diff
@@ -585,6 +618,71 @@ cpdef poolingBackward(
+        <void*>alpha, <void*>beta, <TensorDescriptor>xDesc,
+        <void*>x, <TensorDescriptor>yDesc, <void*>y,
+        <TensorDescriptor>bnScaleBiasMeanVarDesc, <void*>bnScale,
+        <void*>bnBias, <float>exponentialAverageFactor,
```
`<double>exponentialAverageFactor` is correct.

okuta commented on cupy/cuda/cudnn.pxd:
```diff
@@ -235,6 +236,39 @@ cpdef poolingBackward(
+        size_t alpha, size_t beta, size_t xDesc,
+        size_t x, size_t yDesc, size_t y,
+        size_t bnScaleBiasMeanVarDesc, size_t bnScale,
+        size_t bnBias, float exponentialAverageFactor,
```
`double exponentialAverageFactor` is correct.

okuta commented on cupy/cuda/cudnn.pxd:
```diff
@@ -235,6 +236,39 @@ cpdef poolingBackward(
+        size_t bnScaleBiasMeanVarDesc, size_t bnScale,
+        size_t bnBias, float exponentialAverageFactor,
+        size_t resultRunningMean, size_t resultRunningVariance,
+        float epsilon, size_t resultSaveMean, size_t resultSaveInvVariance)
```
`double epsilon` is correct.

okuta commented on cupy/cuda/cudnn.pxd:
```diff
@@ -235,6 +236,39 @@ cpdef poolingBackward(
+        size_t x, size_t yDesc, size_t y,
+        size_t bnScaleBiasMeanVarDesc, size_t bnScale,
+        size_t bnBias, size_t estimatedMean, size_t estimatedVariance,
+        float epsilon)
```
`double epsilon` is correct.

okuta commented on cupy/cuda/cudnn.pxd:
```diff
@@ -235,6 +236,39 @@ cpdef poolingBackward(
+        size_t dy, size_t dxDesc, size_t dx,
+        size_t dBnScaleBiasDesc, size_t bnScale,
+        size_t dBnScaleResult, size_t dBnBiasResult,
+        float epsilon, size_t savedMean, size_t savedInvVariance)
```
`double epsilon` is correct.

okuta commented on cupy/cuda/cudnn.pyx:
```diff
@@ -151,6 +151,39 @@ cdef extern from "cupy_cudnn.h":
+	    void* alpha, void* beta, TensorDescriptor xDesc,
+	    void* x, TensorDescriptor yDesc, void* y,
+	    TensorDescriptor bnScaleBiasMeanVarDesc, void* bnScale,
+	    void* bnBias, float exponentialAverageFactor,
```
Type miss. `double` is correct.

okuta commented on cupy/cuda/cudnn.pyx:
```diff
@@ -151,6 +151,39 @@ cdef extern from "cupy_cudnn.h":
+	    TensorDescriptor bnScaleBiasMeanVarDesc, void* bnScale,
+	    void* bnBias, float exponentialAverageFactor,
+	    void* resultRunningMean, void* resultRunningVariance,
+	    float epsilon, void* resultSaveMean, void* resultSaveInvVariance)
```
Type miss. `double` is correct.

okuta commented on cupy/cuda/cudnn.pyx:
```diff
@@ -151,6 +151,39 @@ cdef extern from "cupy_cudnn.h":
+	    void* x, TensorDescriptor yDesc, void* y,
+	    TensorDescriptor bnScaleBiasMeanVarDesc, void* bnScale,
+	    void* bnBias, void* estimatedMean, void* estimatedVariance,
+	    float epsilon)
```
Type miss. `double` is correct.

okuta commented on cupy/cuda/cudnn.pyx:
```diff
@@ -151,6 +151,39 @@ cdef extern from "cupy_cudnn.h":
+	    TensorDescriptor dxDesc, void* dx,
+	    TensorDescriptor dBnScaleBiasDesc, void* bnScale,
+	    void* dBnScaleResult, void* dBnBiasResult,
+	    float epsilon, void* savedMean, void* savedInvVariance)
```
Type miss. `double` is correct.

okuta commented on cupy/cuda/cudnn.pyx:
```diff
@@ -585,6 +618,71 @@ cpdef poolingBackward(
+        size_t alpha, size_t beta, size_t xDesc,
+        size_t x, size_t yDesc, size_t y,
+        size_t bnScaleBiasMeanVarDesc, size_t bnScale,
+        size_t bnBias, float exponentialAverageFactor,
```
Type miss. `double` is correct.

okuta commented on cupy/cuda/cudnn.pyx:
```diff
@@ -585,6 +618,71 @@ cpdef poolingBackward(
+        size_t bnScaleBiasMeanVarDesc, size_t bnScale,
+        size_t bnBias, float exponentialAverageFactor,
+        size_t resultRunningMean, size_t resultRunningVariance,
+        float epsilon, size_t resultSaveMean, size_t resultSaveInvVariance):
```
Type miss. `double` is correct.

okuta commented on cupy/cuda/cudnn.pyx:
```diff
@@ -585,6 +618,71 @@ cpdef poolingBackward(
+        <void*>x, <TensorDescriptor>yDesc, <void*>y,
+        <TensorDescriptor>bnScaleBiasMeanVarDesc, <void*>bnScale,
+        <void*>bnBias, <void*>estimatedMean, <void*>estimatedVariance,
+	    <float>epsilon)
```
`<float>` is not needed.

okuta commented on cupy/cuda/cudnn.pyx:
```diff
@@ -585,6 +618,71 @@ cpdef poolingBackward(
+	    <TensorDescriptor>dxDesc, <void*>dx,
+	    <TensorDescriptor>dBnScaleBiasDesc, <void*>bnScale,
+	    <void*>dBnScaleResult, <void*>dBnBiasResult,
+	    <float>epsilon, <void*>savedMean, <void*>savedInvVariance)
```
`<float>` is not needed.

okuta commented on cupy/cuda/cudnn.pyx:
```diff
@@ -585,6 +618,71 @@ cpdef poolingBackward(
+        size_t x, size_t yDesc, size_t y,
+        size_t bnScaleBiasMeanVarDesc, size_t bnScale,
+        size_t bnBias, size_t estimatedMean, size_t estimatedVariance,
+        float epsilon):
```
Type miss. `double` is correct.

okuta commented on cupy/cuda/cudnn.pyx:
```diff
@@ -585,6 +618,71 @@ cpdef poolingBackward(
+        size_t dy, size_t dxDesc, size_t dx,
+        size_t dBnScaleBiasDesc, size_t bnScale,
+        size_t dBnScaleResult, size_t dBnBiasResult,
+        float epsilon, size_t savedMean, size_t savedInvVariance):
```
Type miss. `double` is correct.

okuta commented:
Please add with/without cuDNN tests.
See also
https://github.com/pfnet/chainer/blob/master/tests/chainer_tests/functions_tests/activation_tests/test_relu.py#L49
https://github.com/pfnet/chainer/blob/master/tests/chainer_tests/functions_tests/activation_tests/test_relu.py#L57


And, please check calling cuDNN.
See also
https://github.com/pfnet/chainer/blob/master/tests/chainer_tests/functions_tests/activation_tests/test_relu.py#L90

okuta commented:
> I have modified the tests to be similar to the newer batch normalization tests in master that check all float types. float16 and float64 types are tested now. I still do not allow the cuDNN version to use float16 since it fails the test on my machine using cuDNN v5. So if float16 is detected, cupy version is used instead. How does this look?

I tested float16. It fails too...
This PR works to use float32 and float64.

bkvogel commented:
Thank you for the detailed comments. I will make the suggested changes.

bkvogel commented:
I have implemented the requested changes and added additional tests (calling cuDNN with/without cuDNN). I fixed the batch normalization function so that backwards compatibility with existing documented APIs is not broken. Could you check if it looks okay now? Thank you.

okuta commented on tests/chainer_tests/functions_tests/normalization_tests/test_batch_normalization.py:
```diff
@@ -134,20 +149,69 @@ def test_forward_cpu(self):
+        self.var = self.x.var(axis=self.aggr_axes) + self.eps
+        self.expect = self.use_cudnn and (
+            cuda.cudnn.cudnn.getVersion() >= 5000 or
+            self.dtype != numpy.float16)
```
`self.dtype` is not `numpy.float16` always.

okuta commented on chainer/functions/normalization/batch_normalization.py:
```diff
@@ -27,130 +66,294 @@ def check_type_forward(self, in_types):
     """
-    return BatchNormalizationFunction(eps)(x, gamma, beta, mean, var)
+    return BatchNormalizationFunction(eps, mean, var, False,
+                                      0.0, use_cudnn)(x, gamma, beta)
```
The gradients of `mean` and `var` is not calculated.
It is different from previous version.

bkvogel replied:
OK, I will change to back so that `mean` and `var` are passed as in the existing version. I don't see the need to compute the gradients of these (and cuDNN does not compute them and so the revised code will be a bit slower) but I agree it is good not to break compatibility.

okuta replied:
Actually, `links.BatchNormalization` dose not need gradient of `mean` and `var` .
I think reducing extra calculation is OK. For example, adding a control flag.

bkvogel replied:
Yes, since the BN link does not need the gradients of `mean` and `var`, I think it should be possible to maintain backward compatibility of the BN function while also reducing computation by not computing these gradients. I will try this.

bkvogel commented:
I have modified the BN function so that it now computes the gradients of all Variables, maintaining backward compatibility with the existing version. The (slow) code for computing the gradients of `mean' and `var` has been moved to the backward() function (which is not used by the BN link) and so the forward() call will use cuDNN if available and should be fast.
Could you check again if it looks ok now?

okuta commented on chainer/functions/normalization/batch_normalization.py:
```diff
@@ -42,105 +73,289 @@ def forward(self, inputs):
+            # Factor used in the moving average
+            factor = 1 - self.decay
+            # computation runningMean
+            if self.mean_cache is None:
```
How about moving this block into `if self.train` block?
https://github.com/pfnet/chainer/pull/1310/files#diff-148edea535bb806260c487ca31869a91R152

okuta commented on chainer/functions/normalization/batch_normalization.py:
```diff
@@ -42,105 +73,289 @@ def forward(self, inputs):
+                # If cuDNN was called on the forward call, self.x_hat will
+                # be None.
+                if self.x_hat is None:
+                    self.std = xp.sqrt(self.fixed_var,
```
How about writing function to calculate `std` and `x_hat`.
There are similar codes.
https://github.com/pfnet/chainer/pull/1310/files#diff-148edea535bb806260c487ca31869a91R116
https://github.com/pfnet/chainer/pull/1310/files#diff-148edea535bb806260c487ca31869a91R179
https://github.com/pfnet/chainer/pull/1310/files#diff-148edea535bb806260c487ca31869a91R220

okuta commented on chainer/functions/normalization/batch_normalization.py:
```diff
@@ -42,105 +73,289 @@ def forward(self, inputs):
             y = gamma * self.x_hat
             y += beta
+            # cuDNN Batch normalization requires v5 or greater.
+        elif cuda.cudnn_enabled and self.use_cudnn and self.cudnn_dim_ok and \
```
How about changing to following style?
``` python
if (xp is not numpy and cuda.cudnn_enabled and self.use_cudnn and self.cudnn_dim_ok and _cudnn_version >= 5000):
     ~~~~~
else:
    if self.train:
        axis = (0,) + tuple(range(head_ndim, x.ndim))
        mean = x.mean(axis=axis)
        var = x.var(axis=axis)
        var += self.eps
    else:
        mean = self.fixed_mean
        var = self.fixed_var
        self.std = xp.sqrt(var, dtype=var.dtype)
        if xp is numpy:
        ~~~~~
````

bkvogel replied:
Yes, this will reduce code duplication. I will make the change.

okuta commented on chainer/functions/normalization/batch_normalization.py:
```diff
@@ -42,105 +73,289 @@ def forward(self, inputs):
         beta (Variable): The shifting parameter of scaled normalized data.
         eps (float): Epsilon value for numerical stability.
+        running_mean (array): The running average of the mean. This is a
+        running average of the mean over several
```
Please fix an indent.

okuta commented on chainer/functions/normalization/batch_normalization.py:
```diff
@@ -42,105 +73,289 @@ def forward(self, inputs):
         eps (float): Epsilon value for numerical stability.
+        running_mean (array): The running average of the mean. This is a
+        running average of the mean over several
+            mini-batches using the decay parameter. If None, the running
```
Please use ```` ``None`` ````.

okuta commented on chainer/functions/normalization/batch_normalization.py:
```diff
@@ -42,105 +73,289 @@ def forward(self, inputs):
+        running_mean (array): The running average of the mean. This is a
+        running average of the mean over several
+            mini-batches using the decay parameter. If None, the running
+            average is not computed. If this is None,
```
ditto

okuta commented on chainer/functions/normalization/batch_normalization.py:
```diff
@@ -42,105 +73,289 @@ def forward(self, inputs):
+            average is not computed. If this is None,
+            then runnng_var must also be None.
+        running_var (array): The running average of the variance. This is a
+        running average of the variance
```
Please fix an indent.

okuta commented on chainer/functions/normalization/batch_normalization.py:
```diff
@@ -42,105 +73,289 @@ def forward(self, inputs):
+            then runnng_var must also be None.
+        running_var (array): The running average of the variance. This is a
+        running average of the variance
+            over several mini-batches using the decay parameter. If None, the
```
Please use ```` ``None`` ````.

okuta commented on chainer/functions/normalization/batch_normalization.py:
```diff
@@ -42,105 +73,289 @@ def forward(self, inputs):
+        running average of the variance
+            over several mini-batches using the decay parameter. If None, the
+            running average is not computed. If
+            this is None, then running_mean must also be None.
```
ditto

okuta commented on chainer/functions/normalization/batch_normalization.py:
```diff
@@ -42,105 +73,289 @@ def forward(self, inputs):
-    statistics are given by the caller as variables. This is used on testing
-    mode of the batch normalization layer, where batch statistics cannot be
-    used for prediction consistency.
+    statistics are given by the caller as fixed parameters. This is
```
Is this fix needed?

bkvogel replied:
Not needed. I will change wording back to "variables" instead of "parameters."

okuta commented:
How about changing below statement?
https://github.com/pfnet/chainer/pull/1310/files#diff-0fb203b8a923ff41fabe22b3b10508fbR132
`BatchNormalization` do not need gradients of `mean` and `bar`.

okuta commented on chainer/functions/normalization/batch_normalization.py:
```diff
@@ -42,105 +73,289 @@ def forward(self, inputs):
+                self.running_mean = xp.array(self.running_mean)
+                self.running_var = xp.array(self.running_var)
+        elif len(inputs) == 5:
+                self.fixed_mean = xp.array(inputs[3])
```
How about `self.fixed_mean = inputs[3]`?

bkvogel replied:
OK.

okuta commented on chainer/functions/normalization/batch_normalization.py:
```diff
@@ -4,19 +4,56 @@
+def _xhat(x, mean, std, expander):
+    x_mu = x - mean[expander]
+    # Return x_hat
+    return x_mu / std[expander]
```
How about following code?
```python
x_mu /= std[expander] # reduce memory usage
return x_mu
```

okuta commented on chainer/links/normalization/batch_normalization.py:
```diff
@@ -101,7 +100,7 @@ def __call__(self, x, test=False, finetune=False):
         """
         use_batch_mean = not test or finetune
-
+        # use_batch_mean = (not test) or finetune
```
Please remove this line.

okuta commented:
LGTM except comments!

bkvogel commented:
Thank you for the comments. I have made the requested changes.

okuta commented on cupy/cuda/cupy_cudnn.h:
```diff
@@ -118,6 +118,76 @@ cudnnStatus_t cudnnSetConvolutionNdDescriptor_v3(
+cudnnStatus_t cudnnDeriveBNTensorDescriptor(
+         cudnnTensorDescriptor_t derivedBnDesc,
+         const cudnnTensorDescriptor_t xDesc,
+         cudnnBatchNormMode_t mode) {
```
Please define `cudnnBatchNormMode_t`.
Build is failed when CuPy uses lower than cuDNN v5.

okuta commented:
LGTM!

jekbradbury commented:
I'm seeing cuDNN errors when using batchnorm at inference time for 2D tensors, on both cuDNN 5.0 and 5.1.

```
  File "/home/james.bradbury/chainer-dev/chainer/chainer/links/normalization/batch_normalization.py", line 132, in __call__
    x, gamma, beta, mean, var, self.eps)
  File "/home/james.bradbury/chainer-dev/chainer/chainer/functions/normalization/batch_normalization.py", line 347, in fixed_batch_normalization
    use_cudnn)(x, gamma, beta, mean, var)
  File "/home/james.bradbury/chainer-dev/chainer/chainer/function.py", line 130, in __call__
    outputs = self.forward(in_data)
  File "/home/james.bradbury/chainer-dev/chainer/chainer/functions/normalization/batch_normalization.py", line 157, in forward
    self.eps)
  File "cupy/cuda/cudnn.pyx", line 655, in cupy.cuda.cudnn.batchNormalizationForwardInference (cupy/cuda/cudnn.cpp:8636)
  File "cupy/cuda/cudnn.pyx", line 669, in cupy.cuda.cudnn.batchNormalizationForwardInference (cupy/cuda/cudnn.cpp:8439)
  File "cupy/cuda/cudnn.pyx", line 243, in cupy.cuda.cudnn.check_status (cupy/cuda/cudnn.cpp:1507)
cupy.cuda.cudnn.CuDNNError: CUDNN_STATUS_NOT_SUPPORTED: b'CUDNN_STATUS_NOT_SUPPORTED'
```
This is the error I get for both `_cudnn_version=v5005` and `_cudnn_version=v5105`. The cuDNN User Guide for v5.1 says (for all three batchnorm API functions):
> Only 4D and 5D tensors are supported.

And its description of `CUDNN_BATCHNORM_PER_ACTIVATION` (which is for non-convolutional layers, which Chainer assumes are 2D) is
> Normalization is performed per-activation.
> This mode is intended to be used after non-
> convolutional network layers. In this mode bnBias
> and bnScale tensor dimensions are 1xCxHxW.

But this PR explicitly says cuDNN supports 2D tensors:
`self.cudnn_dim_ok = x.ndim == 2 or x.ndim == 4`
Is that true for an older cuDNN version (?), or only at train time (??) or am I doing something wrong...
